### DISCUSSION TITLE
Precise Text Positioning

### DISCUSSION THREAD

**Discussion by FrothyFrog** [fabricjs/fabric.js#9353] [Q&A]
I'm recently new to Fabric and I find there's one thing that really puzzles me. There doesn't seem to be a clean way of precisely positioning text, but then again it may be that it's just as hard to find good documentation. 

The bounding box data returned by a text object using getBoundingRect is clearly off to the point where it can't be used to reliably position the text, based on my understanding.  The area (padding?) around the text within the bounding box varies greatly depending on the font size and angle.  To precisely site the text you simply can't have a variable amount of space within the bounding box on all four sides of the text.  Ideally, it should be a clean cutout, but it's not.

The only thing that seems to give a reliable metric on the text is using getCenterPoint after Left and Top have been set.  This does indeed return the center, as opposed to calculating the center based on the data from GetBoundingRect, which should be usable but it falls far from the mark.

Generally, my technique is to find a move-to point and then move the text bounding box from its center to the move-to point.  But I can't find any way to do this in Fabric, and I've searched for three days.

What is it I'm I missing, or is this just the way it is? Is there a resource that goes into detail on this?

**Comment by ShaMan123** [comment]
Please share a reproduction

**Reply by FrothyFrog** [reply]
Not as easy to do, but I'll get something to you soon.

**Comment by asturur** [comment]
@FrothyFrog the text in fabric ( and every object ) is positioned by the top left corner of the bounding box.
Fabric was built in a moment in which advanced text apis didn't exist and is not leveraging them yet ( and won't for a while yet ).
So what you get is that the text has a bounding box that is used to accomodate the font, an overline, most of the descendants, an underline. Some fonts are also completely out of the box.

It would be nice if you could explain what you are looking for precise text positioning.
There are some magic numbers that given the top left corner ( or any corner ) can let you figure out where the baseline of the font is, but i have to dig them.

**Reply by FrothyFrog** [reply]
Thanks for that info. I think I'm hampered because I can't find meaningful tutorials and documentation so I end up trying to put something together based on work I did in js, php, etc. and it's not working.  I did however discover aCoords by inspecting a text object and it does a much better job of eliminating excess head and foot room such that I believe I can manipulate the positioning after the fact. I still have to create the object to get the text width and height first, then do the calculations adjusting the position of the top left and the angle, then add it to a group, and finally to the canvas.  I should tell you I'm using only numbers so as far as descendants and the like I can treat those areas like padding that I can encroach on.  My application essentially deals with property lines.  When you're working on a plot out in the country there's a lot of space to put things however you like, but when you're working on a small irregular corner lot downtown with crazy polylines I've had to sacrifice consistency by changing font sizes, etc. on the fly.  But I'm encouraged by the support and anticipate getting a solution together when I can work on it tomorrow.  Thanks again.

**Reply by asturur** [reply]
Yes but if you can explain better how do you want to put objects i can help.
What is your flow process to plot etc etc

**Reply by taarss** [reply]
Hello! @asturur I'm experiencing a similar challenge with my program. Once text is positioned, it undergoes serialization to XML and is subsequently utilized by the DevExpress PDF processor. The issue I'm encountering is that all elements are rendered in their exact intended positions except for the texts. I suspect that this problem may be related to the bounding box dimensions of the text elements. Could you kindly share the specific values or "magic numbers" you mentioned earlier?

**Reply by asturur** [reply]
XML as svg?

**Reply by taarss** [reply]
Let me clarify further. My program is designed to create a visual layout or representation. To achieve this, I've developed a custom serializer that transforms fabric elements into a specific XML format. This format is then consumed by another application, which expects this precise XML structure. The transformation process involves converting positional values from pixels to millimeters, and font sizes to points, among other adjustments.

This external application utilizes the XML to accurately overlay elements onto an existing PDF, positioning them exactly as defined within the fabric framework. Although this workflow is not ideal, it is necessary in my scenario due to certain constraints I prefer not to detail.

The challenge arises with the positioning of text elements. Unlike other elements, which are placed precisely as anticipated, text elements deviate from their expected positions. This discrepancy becomes more pronounced with larger font sizes, resulting in a significant shift.

It's important to clarify that the issue does not lie in how the elements are rendered within the fabric interface. Rather, the problem occurs when iterating over all elements on the canvas to retrieve their positions; the positional values for text elements do not align with expectations.

I suspect the root of the problem may be related to the bounding box dimensions of the text elements, as outlined in the discussion.
So my goal would be to somehow calculate where the actual coordinates of the top left of the text element is. As in the actual ink of the text or baseline

I'd very much appreciate if you could tell me what those magic numbers that would let me figure out where the baseline of the font is

**Reply by asturur** [reply]
Ok so, this is a long discussion. I think the first thing we need to figure out is how the text element are rendered in your custom software.
When you export fabricJS text in a new PDF, how do  you represent them? as pdf text elements? as pdf paths? as pdf raster layers?
I have somewhere the code to move from text to path in PDF with exact positioning. That for sure will solve some mistery.
But how are you adding the text into the PDF?

**Reply by taarss** [reply]
My fabric js solution builds on top of an existing solutions that adds elements to the pdf with DevExpress' pdf processor

**Reply by asturur** [reply]
But is text an actual piece of text, or a vector path or an image?

**Reply by taarss** [reply]
its actual pieces of text

**Reply by taarss** [reply]
I've looked into the other software that consumes my xml and this is how it draws the text on pdfs
`        public override void Draw(PdfGraphics graph, double pageWidth, double pageHeight, PdfDocumentProcessor unused)
        {
            CoordinateXY location = RotateText(graph, OriginCalc(pageWidth, pageHeight));

            PointF p = new PointF((float)location.X, (float)location.Y);

            graph.DrawString(Text, Font, PenBrush, p, Format);
            graph.RestoreGraphicsState();
        }`

**Reply by asturur** [reply]
I checked my old code and it was converting text to path and then add the path to PDF.
The conversion of the path was done with [openFont.js](https://opentype.js.org/) that does not know anything about fabric, and then the path was added in the pdf at the text top/left. And it was matching just fine.

this is what i was doing on the PDF;

```js
  fabric.EditorTextBox.prototype.renderPdfKit = async function (pdfDoc, artboard) {
    pdfDoc.save();
    this.transformPdf(pdfDoc); // transformPDF is adding the transformMatrix of the text to the pdf, this is important.
    let lineHeights = 0;
    const left = this._getLeftOffset(); 
    const top = this._getTopOffset();
    this.renderPdfTextLinesBackground(pdfDoc, left, top);
    for (let i = 0, len = this._textLines.length; i < len; i++) { // text is rendered line by line
      const heightOfLine = this.getHeightOfLine(i);
      const maxHeight = heightOfLine / this.lineHeight;
      const leftOffset = this._getLineLeftOffset(i);
      this.renderTextDecorationPdf(pdfDoc, artboard, 'underline', 'stroke');
      this.renderTextDecorationPdf(pdfDoc, artboard, 'underline', 'fill');
      await this.renderLineAndBullet(
        artboard,
        pdfDoc,
        this._textLines[i],
        left + leftOffset,
        top + lineHeights + maxHeight,
        i,
      );
      this.renderTextDecorationPdf(pdfDoc, artboard, 'linethrough', 'fill');
      lineHeights += heightOfLine;
    }
    pdfDoc.restore();
  };
```

So as long as you are using the same transformation logic of fabricJS and are rendering the text line by line you should be ok

**Reply by taarss** [reply]
Okay im not sure i quite understand this solution.
I cant really change how i render the text on the pdf due to maintaining backwords compatability.
I must simply just get the correct coordinates of a text element when im converting all of elements to xml.
Would i be able to use openFont.js to get the correct coordinates that the text will be rendered at? 
If so could calculate the offset that i need to shift my coordinates by like this.
Offset = fabricTextElementCoordinates - pdfRenderedCoordinates

That would solve my problem.

**Reply by asturur** [reply]
You don't have to  use openFont, openFont is just a sort of proof that left and top from fabricJS are good enough.

```
        PointF p = new PointF((float)location.X, (float)location.Y);

        graph.DrawString(Text, Font, PenBrush, p, Format);
        graph.RestoreGraphicsState();
```

instead of render the text at location x and y, you should try to see if your pdf library support canvas transformation with a matrix and then render text at locations described by the code i sent you.
That would mean that you have to export all the returned values of getHeightOfLine(), getLineOffset() for every string of text in your XML and the text object transformMatrix.
